tools/topology/v2/attributes.conf

```diff
@@ -0,0 +1,55 @@
+# common attributes and attribute types. Attributes, defined here can then be
+# used by any blocks. E.g. the "format" attribute can be used by pipelines,
+# DAIs, etc.
+
+# we're using an "availability" parameter in attribute declaration instead of
+# adding ".optional" or ".mandatory" to the array name
+
+DefineType.format {
+	enum [
+		"s16le"
+		"s24le"
+		"s32le"
+		"float"
+	]
+}
+
+DefineType.time_domain {
+	enum [
+		"dma"
+		"timer"
+	]
+}
+
+DefineAttribute.tokens {
+	type array.tuple
+}
+
+DefineAttribute.format {
+	type format
+	availability mandatory
+}
+
+DefineAttribute.template {
+	type string
+	availability mandatory
```

Comment:
yeah, I remember that was your proposal. And I understand the improvement, that it offers, but it would make the code look less readable IMHO. Also, I don't think you can make this work 100%. The same attribute within the same component can be mandatory in some configurations and optional in others, can it not? I don't think you can catch all such errors at compile-time, some can only be really verified at run-time? We could omit availability altogether for now and see later, if you feel strongly about it?